packages/babel-plugin-proposal-pattern-matching/README.md

```diff
@@ -0,0 +1,265 @@
+object in object, object in array, array in object, etc.
+
+## Reducer in Redux
+
+Writing a reducer in redux:
+```js
+function todoApp(state = initialState, action) {
+  switch (action.type) {
+    case SET_VISIBILITY_FILTER:
+      return Object.assign({}, state, {
+        visibilityFilter: action.filter
+      })
+    case ADD_TODO:
+      return Object.assign({}, state, {
+        todos: [
+          ...state.todos,
+          {
+            text: action.text,
+            completed: false
+          }
+        ]
+      })
+    default:
+      return state
+  }
+}
+```
+
+But with pattern matching:
+
+```js
+function todoApp(state = initialState, action) {
+  return match(action) {
+    {type: "SET_VISIBILITY_FILTER", filter}:
+      Object.assign({}, state, {
+        visibilityFilter: filter
+      }),
+    {type: "ADD_TODO", text}:
+      Object.assign({}, state, {
+          todos: [
+            ...state.todos,
+            {
+              text: text,
+              completed: false
+            }
+          ]
+        }),
+    else: state
+  }
+
+```
+
+It's much more cleaner.
+
+## Match rules
+
+### Identifier
+
+You can **only** match identifier in the outermost pattern.
+The identifier should be **only** the following type:
+
+- RegExp variable(Or something contain Symbol.match method), inline RegExp is not supported
+- Array
+- class constructor
+
+Example:
+```js
+var reg = /ok/;
+var some = "that";
+
+function Foo(name) {
+  this.name = name
+}
+
+var foo = new Foo("foo");
+
+match("that") {
+  reg: "will call reg[Symbol.match] method to test",
+  Array: "this is an array",
+  foo: "this is an instance of Foo",
+  else: "nothing"
+}
+```
+
+### Array
+
+When matching array, it will check if it's array,
+then check the length of array if the pattern does not
+contain `...`. Last, match the pattern in the array.
+
+Example:
+```js
+// pattern
+[1,2,3]
+// generating code
+Array.isArray(_match_expr) &&
+_match_expr.length === 3 &&
+_match_expr[0] === 1 &&
+_match_expr[1] === 2 &&
+_match_expr[2] === 3
+```
+
+And rest identifier is supported.
+
+Example:
+
+```js
+// rest elements will be added to `rest`
+[1,2,3, ...rest]
+```
+
+### Object
+
+Matching object is more complicated, it will check
+property of object. If the object does not have the
+property key, it will fail.
+
+Example:
+```js
+// pattern
+{a, b: "b", ...c}
+// generating code
+typeof _match_expr === "object" &&
+_match_expr.hasOwnProperty("a") &&
+_match_expr.a !== void 0 &&
+_match_expr.hasOwnProperty("b") &&
+_match_expr.b === "b"
+```
+
+The variable `c` is a rest pattern identifier,
+the properties which haven't matched will be
+bind to the object `c`. It's obvious that `c`
+is a new object.
+
+### Else
+
+The `else` pattern can match all the cases.
+All the clauses after `else` **will be abandoned**.
+
+Example:
+```js
+match (4) {
+  1: "one",
+  2: "two",
+  3: "three",
+  else: "else",
+  4: "abandoned"
+}
+```
```

Comment:
This is spec compliant but it makes no sens to me to allow unreachable clauses after the else.

Reply:
@xtuc I agree with you. See the latest commit.

Reply:
Actually since it was spec compliant we should probably keep it. I opened an issue for that let's see.

Reply:
@xtuc OK